### PR TITLE
sarkar/Fix TRL training issues

### PR DESCRIPTION
**Pull request by ssarkar2**
# What does this PR do?
Avoid nan issue in TRL by bypassing fusedrope
Fix compute_rewards outputs in ppo_trainer.py

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Review by ssarkar2 — COMMENTED** [review]
add a TODO in this line to use model.config instead of model.generation_config. 

https://github.com/huggingface/optimum-habana/blob/ba43970c89083e0bf252c28379efac27a47cef03/examples/language-modeling/run_lora_clm.py#L625

**Review by ssarkar2 — COMMENTED** [review]
Should we not add use_fused_rope to configuration_llama.py?

we can just set model.config.use_fused_rope, but having it in config_llama.py gives some indicatin that it is an useable flag

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 1]
```diff
@@ -772,6 +772,12 @@ def forward(
             batch_size, seq_length = inputs_embeds.shape[:2]
         else:
             raise ValueError("You have to specify either input_ids or inputs_embeds")
+        if hasattr(self.config, "use_fused_rope") and self.config.use_fused_rope is False:
+            global has_fused_rope
+            has_fused_rope = False
+        if hasattr(self.config, "use_fused_rms_norm") and self.config.use_fused_rms_norm is False:
+            global has_fused_rms_norm
+            has_fused_rms_norm = False
```
Comment (ssarkar2) [inline optimum/habana/transformers/models/llama/modeling_llama.py:780]: has_fused_rms_norm also has some issues?

atleast in sft, dpo, ppo i see only use_fused_rope flag set to false

File: optimum/habana/transformers/models/llama/configuration_llama.py [hunk 2]
```diff
@@ -26,6 +26,7 @@ def __init__(
         attention_bias=False,
         attention_dropout=0.0,
         mlp_bias=False,
+        use_fused_rope=False,
```
[the remark this answers is not in the document]
Comment (sywangyi) [reply optimum/habana/transformers/models/llama/configuration_llama.py]: no need to add this, kwargs is supported

File: optimum/habana/transformers/models/llama/modeling_llama.py (hunk 1 above)
Comment (sywangyi) [reply optimum/habana/transformers/models/llama/modeling_llama.py:780]: ppo has backward issue in rms. do not see in dpo and sft.

File: optimum/habana/trl/trainer/ppo_trainer.py [hunk 3]
```diff
@@ -539,11 +539,11 @@ def step(
                 active_full_logprobs = logprobs_from_logits(logits_or_none, None, gather=False)
                 ref_full_logprobs = logprobs_from_logits(ref_logits_or_none, None, gather=False)
 
-                rewards, non_score_reward = self.compute_rewards(
+                rewards, non_score_reward, kls = self.compute_rewards(
                     scores, active_full_logprobs, ref_full_logprobs, masks
                 )
             else:
-                rewards, non_score_reward = self.compute_rewards(scores, all_logprobs, ref_logprobs, masks)
+                rewards, non_score_reward, kls = self.compute_rewards(scores, all_logprobs, ref_logprobs, masks)
```
Comment (ssarkar2) [inline optimum/habana/trl/trainer/ppo_trainer.py:546]: @libinta , how was this not caught in tests earlier, dont we have ppo tests?

Comment (sywangyi) [reply optimum/habana/trl/trainer/ppo_trainer.py:546]: ppo ci is here. https://github.com/huggingface/optimum-habana/pull/1020. after this PR is merged. I need to update 1020, please help merge it then.

**Comment by sywangyi** [comment]
> Should we not add use_fused_rope to configuration_llama.py?
> 
> we can just set model.config.use_fused_rope, but having it in config_llama.py gives some indicatin that it is an useable flag

I have removed it from configuration_llama.py

File: examples/trl/sft.py [hunk 4]
```diff
@@ -151,6 +151,7 @@ def create_datasets(tokenizer, args, seed=None):
     token=script_args.token,
 )
 base_model.config.use_cache = False
+base_model.config.use_fused_rope = False
```
Comment (mfuntowicz) [inline examples/trl/sft.py:154]: No need for `model.config.use_fused_rms_norm = False` here ?

File: examples/trl/reward_modeling.py [hunk 5]
```diff
@@ -181,6 +181,7 @@ class ScriptArguments:
 tokenizer.pad_token = tokenizer.eos_token
 model.config.pad_token_id = tokenizer.eos_token_id
 model.config.use_cache = not script_args.gradient_checkpointing
+model.config.use_fused_rope = False
```
Comment (mfuntowicz) [inline examples/trl/reward_modeling.py:184]: No need for `model.config.use_fused_rms_norm = False` here?

File: optimum/habana/trl/trainer/ppo_trainer.py (hunk 3 above)
Comment (mfuntowicz) [inline optimum/habana/trl/trainer/ppo_trainer.py:546]: It seems unrelated to the current PR, do you want to proceed forward this way? Or do we need another PR for this especially?

File: examples/trl/reward_modeling.py (hunk 5 above)
Comment (ssarkar2) [reply examples/trl/reward_modeling.py:184]: the condition checked is:
if hasattr(self.config, "use_fused_rms_norm") and self.config.use_fused_rms_norm is False:

only 1 place adds use_fused_rms_norm=False (ppo.py). For rest of them hasattr(self.config, "use_fused_rms_norm")  will fail and use_fused_rms_norm wont be set to False

This is really just a way to disable fused_rms in ppo

File: optimum/habana/trl/trainer/ppo_trainer.py (hunk 3 above)
Comment (ssarkar2) [reply optimum/habana/trl/trainer/ppo_trainer.py:546]: This PR we are trying to fix all the stuff in TRL folder, so there are 2 issues: nan and then this output of compute_rewards.

changed the description to reflect this better. I think we can keep both of these in 1 PR

**Review by mfuntowicz — APPROVED** [review]
LGTM !

**Review by regisss — APPROVED** [review]
